runtime/src/iree-amd-aie/driver/hsa/hsa_driver.cc

```diff
@@ -0,0 +1,355 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cstdint>
+#include <cstring>
+#include <vector>
+
+#include "iree-amd-aie/driver/hsa/api.h"
+#include "iree-amd-aie/driver/hsa/dynamic_symbols.h"
+#include "iree-amd-aie/driver/hsa/hsa_device.h"
+#include "iree-amd-aie/driver/hsa/hsa_headers.h"
+#include "iree-amd-aie/driver/hsa/util.h"
+#include "iree/base/api.h"
+#include "iree/base/assert.h"
+#include "iree/base/tracing.h"
+#include "iree/hal/api.h"
+#include "status_util.h"
+
+#define IREE_HAL_HSA_MAX_DEVICE_NAME_LENGTH 64
+#define IREE_HAL_HSA_MAX_DEVICES 64
+#define IREE_HAL_HSA_DEVICE_NOT_FOUND IREE_HAL_HSA_MAX_DEVICES
+
+#define IREE_DEVICE_ID_TO_HSADEVICE(device_id) (int)((device_id) - 1)
+
+constexpr int DEFAULT_DEVICE_INDEX = 0;
+
+struct iree_hal_hsa_driver_t {
+  // Abstract resource used for injecting reference counting and vtable;
+  // must be at offset 0.
+  iree_hal_resource_t resource{};
+  iree_allocator_t host_allocator{};
+  iree_string_view_t identifier{};
+  iree_hal_hsa_dynamic_symbols_t hsa_symbols{};
+  int num_aie_agents{};
+  hsa::hsa_agent_t agents[IREE_HAL_HSA_MAX_DEVICES]{};
```

Comment:
I know you currently do not care to much about the GPU side, but any thoughts if what I originally had made sense?